src/tracks/ui/AffordanceHandle.cpp

```diff
@@ -0,0 +1,105 @@
+
+#include "../../HitTestResult.h"
+#include "../../ProjectAudioIO.h"
+#include "../../RefreshCode.h"
+#include "../../ViewInfo.h"
+#include "../../SelectionState.h"
+#include "../../ProjectSettings.h"
+#include "../../TrackPanelMouseEvent.h"
+#include "../../WaveClip.h"
+#include "../../ProjectHistory.h"
+#include "../../Track.h"
+#include "../../WaveTrack.h"
+#include "../../../images/Cursors.h"
+
+UIHandlePtr AffordanceHandle::HitAnywhere(std::weak_ptr<AffordanceHandle>& holder, const std::shared_ptr<Track>& pTrack)
+    {
+        auto result = std::make_shared<AffordanceHandle>(pTrack);
+        result = AssignUIHandlePtr(holder, result);
+        return result;
+    }
+
+HitTestPreview AffordanceHandle::HitPreview(const AudacityProject* WXUNUSED(pProject), bool unsafe, bool moving)
+{
+    static auto disabledCursor =
+        MakeCursor(wxCURSOR_NO_ENTRY, DisabledCursorXpm, 16, 16);
+    static auto handOpenCursor =
+        MakeCursor(wxCURSOR_HAND, RearrangeCursorXpm, 16, 16);
+    static auto handClosedCursor =
+        MakeCursor(wxCURSOR_HAND, RearrangingCursorXpm, 16, 16);
+    // i18n-hint Appears on hovering mouse over clip affordance
+    auto message = XO("Click and drag to move a clip in time, or click to select");
+
+    if (unsafe)
+        return { message, &*disabledCursor };
+    return {
+        message,
+        (moving
+        ? &*handClosedCursor
+        : &*handOpenCursor)
+    };
+}
+
+HitTestPreview AffordanceHandle::Preview(const TrackPanelMouseState& mouseState, AudacityProject* pProject)
+{
+    const bool unsafe = ProjectAudioIO::Get(*pProject).IsAudioActive();
+    return HitPreview(pProject, unsafe, Clicked());
+}
+
+AffordanceHandle::AffordanceHandle(const std::shared_ptr<Track> track)
+    : TimeShiftHandle(track, false)
+{
+    SetChangeHighlight(RefreshCode::RefreshCell | RefreshCode::RefreshLatestCell);
+}
+
+UIHandle::Result AffordanceHandle::Click(const TrackPanelMouseEvent& evt, AudacityProject* pProject)
+{
+    auto result = TimeShiftHandle::Click(evt, pProject);
+    return result | RefreshCode::RefreshCell;
+}
+
+UIHandle::Result AffordanceHandle::Release(const TrackPanelMouseEvent& event, AudacityProject* pProject, wxWindow* pParent)
+{
+    auto result = TimeShiftHandle::Release(event, pProject, pParent);
+    if (result == RefreshCode::RefreshNone)
+    {
+        //almost the same behaviour as provided by SelectHandle
+        auto& viewInfo = ViewInfo::Get(*pProject);
+        const auto& settings = ProjectSettings::Get(*pProject);
+
+        const auto sTrack = TrackList::Get(*pProject).Lock(std::weak_ptr<Track>(GetTrack()));
+        const auto pTrack = sTrack.get();
+
+        auto& selectionState = SelectionState::Get(*pProject);
+
+        auto& trackList = TrackList::Get(*pProject);
+
+        // Deselect all other tracks and select this one.
+        selectionState.SelectNone(trackList);
+        selectionState.SelectTrack(*pTrack, true, true);
+
+        pTrack->TypeSwitch(
```

Comment:
I don't like track the existing TypeSwitches, I hope to eliminate many of them, and I want to avoid adding more of them.

Better:  AffordanceHandle defines another virtual function;  and WaveTrackAffordanceHandle is a further subclass supplying the behavior in the block below in an override.

Reply:
This is still unresolved

Reply:
Now it's resolved, thanks!

I see you copied the implementation of SelectHandle's double click, which still has one of the unfortunate TypeSwitches, but that can be fixed later.